packages/dso-toolkit/gulp/build-svg-spritesheets.js

```diff
@@ -0,0 +1,165 @@
+import { parse } from "css";
+import svgstore from "gulp-svgstore";
+import cheerio from "gulp-cheerio";
+import prettier from "gulp-prettier";
+import rename from "gulp-rename";
+
+import { sassTransformer } from "./transformers/sass.transformer.js";
+
+const iconsPath = "src/icons";
+const distPath = "dist";
+const outPutNameIcons = "dso-icons.svg";
+const outPutNameIconsWithVariants = "di.svg";
+const canvasSize = 32;
+const gutterSize = 10;
+
+async function parseStylesheets() {
+  return new Promise((resolve, reject) => {
+    gulp
+      .src(`${iconsPath}/*.scss`) // Read all SCSS files in the icons directory
+      .pipe(sassTransformer()) // Transform SCSS into plain CSS
+      .pipe(
+        obj((error, data) => {
+          if (error) {
+            return reject(error);
+          }
+
+          // Process each transformed SCSS file
+          const parsedStyles = data.map(({ relative, extname, contents }) => {
+            const id = basename(relative, extname); // Extract the file name without the extension
+            const style = contents.toString();
+            const ast = parse(style); // Parse the CSS into an abstract syntax tree (AST)
+
+            // Extract color variants for each CSS rule
+            const variants = ast.stylesheet.rules
+              .filter((rule) => rule.type === "rule") // Keep only CSS rules (ignore comments, etc.)
+              .flatMap((rule) => {
+                const colorDeclaration = rule.declarations.find((d) => d.property === "color");
+                if (!colorDeclaration) return [];
+
+                const color = colorDeclaration.value;
+                return rule.selectors
+                  .map((selector) => {
+                    if (selector === id) {
+                      return { selector: "default", color }; // Mark as default variant
+                    }
+
+                    if (selector.startsWith(`${id}:`)) {
+                      return { selector: selector.slice(id.length + 1), color }; // Extract the variant name
+                    }
+
+                    return null;
+                  })
+                  .filter(Boolean); // Remove null entries
+              });
+
+            return { id, variants };
+          });
+
+          resolve(parsedStyles);
+        }),
+      );
+  });
+}
+
+async function buildDiSpritesheet() {
+  const iconStylesheets = await parseStylesheets(); // Parse stylesheets within the function
+
+  if (!iconStylesheets || iconStylesheets.length === 0) {
+    throw new Error("No icon stylesheets found. Pipeline terminated.");
+  }
+
+  return gulp
+    .src(`${iconsPath}/*.svg`)
+    .pipe(
+      svgstore({
+        inlineSvg: true,
+      }),
+    )
+    .pipe(
+      cheerio({
+        run($) {
+          const symbols = $("symbol").toArray();
+          let position = 0;
+
+          // Create a map for quick stylesheet lookup
+          const stylesheetMap = Object.fromEntries(iconStylesheets.map((s) => [s.id, s]));
```

Comment:
Het enige bestaansrecht `const iconStylesheets` (en `parseStylesheets()`) is deze situatie: Waarom geeft `parseStylesheets()` niet direct een format terug waar makkelijk mee gewerkt kan worden?

Overigens is "makkelijk" persoonlijk en is het de prijs waard: extra code voor een nieuw format (nieuwe array) voor index selector ipv. `.find()` ?

Alles wat wij hier doen is snelheid secundair. Met onze lijstjes gaan we het verschil niet merken. Mijn primaire focus is de leesbare code.